File: src/lib/utils/user-default-settings.ts
import type { NewUserDefaultSettingsPayload } from '$lib/apis/users';

type UserDefaultUiTemplate = {
	models: string[];
	showFeaturedAssistantsOnHome: boolean;
	showChatTitleInTab: boolean;
	chatBubble: boolean;
	showUsername: boolean;
	widescreenMode: boolean;
	notificationSound: boolean;
	enableAutoScrollOnStreaming: boolean;
	richTextInput: boolean;
	promptAutocomplete: boolean;
	showFormattingToolbar: boolean;
	insertPromptAsRichText: boolean;
	largeTextAsFile: boolean;
	copyFormatted: boolean;
	copyFormattedUserSet: boolean;
	ctrlEnterToSend: boolean;
	system: string;
	title: { auto: boolean };
	autoTags: boolean;
	autoFollowUps: boolean;
	detectArtifacts: boolean;
	svgPreviewAutoOpen: boolean;
	responseAutoCopy: boolean;
	temporaryChatByDefault: boolean;
	newChatInheritsPreviousState: boolean;
	collapseCodeBlocks: boolean;
	collapseHistoricalLongResponses: boolean;
	showInlineCitations: boolean;
	showMessageOutline: boolean;
	expandDetails: boolean;
	insertSuggestionPrompt: boolean;
	keepFollowUpPrompts: boolean;
	insertFollowUpPrompt: boolean;
	displayMultiModelResponsesInTabs: boolean;
	showFloatingActionButtons: boolean;
	floatingActionButtons:
		| {
				id: string;
				label: string;
				input: boolean;
				prompt: string;
		  }[]
		| null;
};

export type UserDefaultUiBoolKey = {
	[K in keyof UserDefaultUiTemplate]: UserDefaultUiTemplate[K] extends boolean ? K : never;
}[keyof UserDefaultUiTemplate];

export const DEFAULT_USER_DEFAULT_UI_TEMPLATE: UserDefaultUiTemplate = {
	models: [],
	showFeaturedAssistantsOnHome: true,
	showChatTitleInTab: true,
	chatBubble: true,
	showUsername: false,
	widescreenMode: false,
	notificationSound: true,
	enableAutoScrollOnStreaming: true,
	richTextInput: true,
	promptAutocomplete: false,
	showFormattingToolbar: false,
	insertPromptAsRichText: false,
	largeTextAsFile: false,
	copyFormatted: true,
	copyFormattedUserSet: false,
	ctrlEnterToSend: false,
	system: '',
	title: { auto: true },
	autoTags: true,
	autoFollowUps: true,
	detectArtifacts: true,
	svgPreviewAutoOpen: true,
	responseAutoCopy: false,
	temporaryChatByDefault: false,
	newChatInheritsPreviousState: false,
	collapseCodeBlocks: false,
	collapseHistoricalLongResponses: true,
	showInlineCitations: true,
	showMessageOutline: true,
	expandDetails: false,
	insertSuggestionPrompt: false,
	keepFollowUpPrompts: false,
	insertFollowUpPrompt: false,
	displayMultiModelResponsesInTabs: false,
	showFloatingActionButtons: true,
	floatingActionButtons: null
};

const UI_BOOL_KEYS: UserDefaultUiBoolKey[] = [
	'showFeaturedAssistantsOnHome',
	'showChatTitleInTab',
	'chatBubble',
	'showUsername',
	'widescreenMode',
	'notificationSound',
	'enableAutoScrollOnStreaming',
	'richTextInput',
	'promptAutocomplete',
	'showFormattingToolbar',
	'insertPromptAsRichText',
	'largeTextAsFile',
	'copyFormatted',
	'copyFormattedUserSet',
	'ctrlEnterToSend',
	'autoTags',
	'autoFollowUps',
	'detectArtifacts',
	'svgPreviewAutoOpen',
	'responseAutoCopy',
	'temporaryChatByDefault',
	'newChatInheritsPreviousState',
	'collapseCodeBlocks',
	'collapseHistoricalLongResponses',
	'showInlineCitations',
	'showMessageOutline',
	'expandDetails',
	'insertSuggestionPrompt',
	'keepFollowUpPrompts',
	'insertFollowUpPrompt',
	'displayMultiModelResponsesInTabs',
	'showFloatingActionButtons'
];

const UI_STRING_KEYS = ['system'];

const UI_ARRAY_KEYS = ['models'];

const clone = <T>(value: T): T => JSON.parse(JSON.stringify(value));
const asRecord = (value: unknown): Record<string, any> =>
	value && typeof value === 'object' && !Array.isArray(value) ? (value as Record<string, any>) : {};

const isEqual = (a: unknown, b: unknown) => JSON.stringify(a) === JSON.stringify(b);

export const createEmptyNewUserDefaultSettings = (): NewUserDefaultSettingsPayload => ({
	configured: false,
	enabled: false,
	roles: ['user', 'pending'],
	ui: {},
	tools: { native_tools: {} }
});

export const normalizeNewUserDefaultSettings = (
	value: Partial<NewUserDefaultSettingsPayload> | null | undefined
) => {
	const raw = asRecord(value);
	const rawUi = asRecord(raw.ui);
	const defaults = clone(DEFAULT_USER_DEFAULT_UI_TEMPLATE);

	return {
		configured: Boolean(raw.configured),
		enabled: Boolean(raw.enabled),
		roles: Array.isArray(raw.roles)
			? raw.roles.filter((role: unknown) => role === 'user' || role === 'pending')
			: ['user', 'pending'],
		ui: {
			...defaults,
			...rawUi,
			models: Array.isArray(rawUi.models) ? rawUi.models : [],
			title: {
				...defaults.title,
				...asRecord(rawUi.title)
			}
		},
		tools: { native_tools: {} }
	};
};

export const pickUserDefaultUiFields = (ui: Record<string, any>) => {
	const source = asRecord(ui);
	const picked: Record<string, any> = {};

	for (const key of UI_BOOL_KEYS) {
		if (typeof source[key] === 'boolean') picked[key] = source[key];
	}

	for (const key of UI_STRING_KEYS) {
		if (typeof source[key] === 'string') picked[key] = source[key];
	}

	for (const key of UI_ARRAY_KEYS) {
		if (Array.isArray(source[key]))
			picked[key] = source[key].filter((item) => typeof item === 'string');
	}

	if (source.title && typeof source.title.auto === 'boolean') {
		picked.title = { auto: source.title.auto };
	}

	if ('floatingActionButtons' in source) {
		picked.floatingActionButtons =
			source.floatingActionButtons === null
				? null
				: Array.isArray(source.floatingActionButtons)
					? source.floatingActionButtons
					: null;
	}

	return picked;
};

export const buildNewUserDefaultSettingsPayload = (
	value: ReturnType<typeof normalizeNewUserDefaultSettings>
): NewUserDefaultSettingsPayload => {
	const pickedUi = pickUserDefaultUiFields(value.ui);
	const defaultUi = pickUserDefaultUiFields(clone(DEFAULT_USER_DEFAULT_UI_TEMPLATE));
	const ui: Record<string, any> = {};

	for (const [key, nextValue] of Object.entries(pickedUi)) {
		if (!isEqual(nextValue, defaultUi[key])) {
			ui[key] = nextValue;
		}
	}

	if ('copyFormatted' in ui) {
		ui.copyFormattedUserSet = true;
	}

	return {
		configured: Boolean(value.configured),
		enabled: Boolean(value.enabled),
		roles: value.roles.filter((role) => role === 'user' || role === 'pending'),
		ui,
		tools: { native_tools: {} }
	};
};

File: backend/open_webui/utils/user_default_settings.py
import copy
from typing import Any


DEFAULT_NEW_USER_DEFAULT_SETTINGS = {
    "configured": False,
    "enabled": False,
    "roles": ["user", "pending"],
    "ui": {},
    "tools": {"native_tools": {}},
}

ALLOWED_UI_BOOL_KEYS = {
    "showFeaturedAssistantsOnHome",
    "showChatTitleInTab",
    "chatBubble",
    "showUsername",
    "widescreenMode",
    "notificationSound",
    "enableAutoScrollOnStreaming",
    "richTextInput",
    "promptAutocomplete",
    "showFormattingToolbar",
    "insertPromptAsRichText",
    "largeTextAsFile",
    "copyFormatted",
    "copyFormattedUserSet",
    "ctrlEnterToSend",
    "autoTags",
    "autoFollowUps",
    "detectArtifacts",
    "svgPreviewAutoOpen",
    "responseAutoCopy",
    "temporaryChatByDefault",
    "newChatInheritsPreviousState",
    "collapseCodeBlocks",
    "collapseHistoricalLongResponses",
    "showInlineCitations",
    "showMessageOutline",
    "expandDetails",
    "insertSuggestionPrompt",
    "keepFollowUpPrompts",
    "insertFollowUpPrompt",
    "displayMultiModelResponsesInTabs",
    "showFloatingActionButtons",
}

ALLOWED_UI_STRING_KEYS = {
    "system": 12000,
}

ALLOWED_STRING_ARRAY_KEYS = {
    "models",
}


def _as_dict(value: Any) -> dict:
    return value if isinstance(value, dict) else {}


def _is_bool(value: Any) -> bool:
    return isinstance(value, bool)


def _clean_string(value: Any, *, max_length: int = 400) -> str | None:
    if not isinstance(value, str):
        return None
    return value[:max_length]


def _clean_string_array(value: Any, *, max_items: int = 200, max_length: int = 400):
    if not isinstance(value, list):
        return None

    cleaned = []
    for item in value[:max_items]:
        if isinstance(item, str):
            cleaned.append(item[:max_length])
    return cleaned


def _sanitize_title(value: Any) -> dict:
    raw = _as_dict(value)
    cleaned = {}
    if _is_bool(raw.get("auto")):
        cleaned["auto"] = raw["auto"]
    return cleaned


def _sanitize_floating_action_buttons(value: Any):
    if value is None:
        return None
    if not isinstance(value, list):
        return None

    cleaned = []
    for item in value[:20]:
        raw = _as_dict(item)
        button = {
            "id": _clean_string(raw.get("id"), max_length=80),
            "label": _clean_string(raw.get("label"), max_length=80),
            "input": raw.get("input"),
            "prompt": _clean_string(raw.get("prompt"), max_length=8000),
        }
        if (
            button["id"]
            and button["label"]
            and _is_bool(button["input"])
            and button["prompt"] is not None
        ):
            cleaned.append(button)

    return cleaned


def sanitize_user_default_ui(value: Any) -> dict:
    raw = _as_dict(value)
    cleaned: dict[str, Any] = {}

    for key in ALLOWED_UI_BOOL_KEYS:
        if _is_bool(raw.get(key)):
            cleaned[key] = raw[key]

    for key, max_length in ALLOWED_UI_STRING_KEYS.items():
        next_value = _clean_string(raw.get(key), max_length=max_length)
        if next_value is None:
            continue
        cleaned[key] = next_value

    for key in ALLOWED_STRING_ARRAY_KEYS:
        if key in raw:
            cleaned[key] = _clean_string_array(raw.get(key)) or []

    title = _sanitize_title(raw.get("title"))
    if title:
        cleaned["title"] = title

    if "floatingActionButtons" in raw:
        buttons = _sanitize_floating_action_buttons(raw.get("floatingActionButtons"))
        if buttons is not None:
            cleaned["floatingActionButtons"] = buttons

    return cleaned


def sanitize_new_user_default_settings(value: Any) -> dict:
    raw = _as_dict(value)
    ui = sanitize_user_default_ui(raw.get("ui"))

    configured = (_is_bool(raw.get("configured")) and raw["configured"]) or bool(ui)

    return {
        "configured": configured,
        "enabled": bool(ui),
        "roles": copy.deepcopy(DEFAULT_NEW_USER_DEFAULT_SETTINGS["roles"]),
        "ui": ui,
        "tools": {"native_tools": {}},
    }


def build_new_user_settings_from_template(value: Any, role: str) -> dict | None:
    template = sanitize_new_user_default_settings(value)

    if not template["enabled"] or role not in template["roles"]:
        return None

    settings: dict[str, Any] = {}
    ui = _as_dict(template.get("ui"))

    if ui:
        settings["ui"] = copy.deepcopy(ui)

    if not settings:
        return None

    settings["revision"] = 0
    return settings
